pinot-server/src/main/java/org/apache/pinot/server/starter/helix/IngestionBasedConsumptionStatusChecker.java

```diff
@@ -19,80 +19,107 @@
+  // This might be called by multiple threads, thus synchronized to be correct.
+  public synchronized int getNumConsumingSegmentsNotReachedIngestionCriteria() {
+    Set<String> tablesWithMissingSegment = new HashSet<>();
+    for (Map.Entry<String, Set<String>> tableSegments : _consumingSegmentsByTable.entrySet()) {
+      String tableNameWithType = tableSegments.getKey();
+      TableDataManager tableDataManager = _instanceDataManager.getTableDataManager(tableNameWithType);
       if (tableDataManager == null) {
-        _logger.info("TableDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        _logger.info("No tableDataManager for table: {}. Will check consumption status later", tableNameWithType);
+        tablesWithMissingSegment.add(tableNameWithType);
         continue;
       }
-      SegmentDataManager segmentDataManager = null;
-      try {
-        segmentDataManager = tableDataManager.acquireSegment(segName);
-        if (segmentDataManager == null) {
-          _logger.info("SegmentDataManager is not yet setup for segment {}. Will check consumption status later",
-              segName);
+      Set<String> consumingSegments = tableSegments.getValue();
+      Set<String> caughtUpSegments = _caughtUpSegmentsByTable.computeIfAbsent(tableNameWithType, k -> new HashSet<>());
+      for (String segName : consumingSegments) {
+        if (caughtUpSegments.contains(segName)) {
           continue;
         }
-        if (!(segmentDataManager instanceof RealtimeSegmentDataManager)) {
-          // There's a possibility that a consuming segment has converted to a committed segment. If that's the case,
-          // segment data manager will not be of type RealtimeSegmentDataManager.
-          _logger.info("Segment {} is already committed and is considered caught up.", segName);
-          _caughtUpSegments.add(segName);
-          continue;
-        }
-
-        RealtimeSegmentDataManager rtSegmentDataManager = (RealtimeSegmentDataManager) segmentDataManager;
-        if (isSegmentCaughtUp(segName, rtSegmentDataManager)) {
-          _caughtUpSegments.add(segName);
+        SegmentDataManager segmentDataManager = null;
+        try {
+          segmentDataManager = tableDataManager.acquireSegment(segName);
+          if (segmentDataManager == null) {
+            _logger.info("No SegmentDataManager for segment: {}. Will check consumption status later", segName);
+            tablesWithMissingSegment.add(tableNameWithType);
+            continue;
+          }
+          if (!(segmentDataManager instanceof RealtimeSegmentDataManager)) {
+            // There's a possibility that a consuming segment has converted to a committed segment. If that's the case,
+            // segment data manager will not be of type RealtimeSegmentDataManager.
+            _logger.info("Segment: {} is already committed and is considered caught up.", segName);
```

Comment:
In this scenario, a new consuming segment should already be created. Consider re-fetch the consuming segment for the table